File: runtime/c/stubs.c
/* Plain - the three tables a program writes for itself.
 *
 * runtime/c/plain.c leans on these, and the translator writes a real one of
 * each for every program. These empty ones exist so the runtime can be
 * built and checked on its own, which the test suite does.
 */

static void plain_defaults(const char *kind, Value into) {
    (void)kind;
    (void)into;
}

static const char *plain_base(const char *kind) {
    (void)kind;
    return NULL;
}

static int plain_do(const char *kind, const char *action, Value me, Value *args, int count, Value *answer) {
    (void)kind;
    (void)action;
    (void)me;
    (void)args;
    (void)count;
    (void)answer;
    return 0;
}

int main(void) {
    PlainFrame frame;
    PlainCatch spot;
    Value list = plain_nothing();
    plain_enter(&frame);
    plain_local(&frame, &list);

    /* Enough to prove the runtime holds together and agrees with Plain. */
    plain_say(plain_add(plain_number_value(0.1), plain_number_value(0.2)));
    plain_say(plain_divide(plain_number_value(10), plain_number_value(4)));

    plain_keep(&list, plain_list_of(3, plain_number_value(3), plain_number_value(1), plain_number_value(2)));
    plain_say(plain_sorted(list));
    plain_say(plain_item(list, plain_number_value(1)));
    plain_say(plain_join_with(list, plain_text_value("-")));
    plain_say(plain_part(plain_text_value("Hello, World"), plain_number_value(1), plain_number_value(5)));

    plain_push_catch(&spot);
    if (PLAIN_SETJMP(spot.jump) == 0) {
        plain_divide(plain_number_value(1), plain_number_value(0));
        plain_pop_catch();
    } else {
        plain_pop_catch();
        plain_say(plain_text_value(spot.words));
    }

    plain_sweep(&frame);
    plain_leave(&frame, plain_nothing());
    plain_done();
    return 0;
}

File: runtime/rust/stubs.rs
// Plain - the three tables a program writes for itself.
//
// runtime/rust/plain.rs leans on these, and the translator writes a real one
// of each for every program. These empty ones exist so the runtime can be
// compiled and checked on its own, which the test suite does.

fn plain_defaults(_kind: &str, _into: &mut Thing) {}

fn plain_base(_kind: &str) -> Option<String> {
    None
}

fn plain_do(_kind: &str, _action: &str, _me: Value, _args: &[Value]) -> Option<Value> {
    None
}

fn main() {
    // Enough to prove the runtime holds together and agrees with Plain.
    println!("{}", plain_show(&plain_add(Value::Number(0.1), Value::Number(0.2))));
    println!("{}", plain_show(&plain_divide(Value::Number(10.0), Value::Number(4.0))));
    let list = plain_list_value(vec![Value::Number(3.0), Value::Number(1.0), Value::Number(2.0)]);
    println!("{}", plain_show(&plain_sorted(list.clone())));
    println!("{}", plain_show(&plain_item(list, Value::Number(1.0))));
    let problem = plain_try(|| {
        plain_divide(Value::Number(1.0), Value::Number(0.0));
    });
    println!("{}", problem.unwrap_or_default());
}
